lib/babe/verify.go

```diff
@@ -109,6 +112,54 @@ func (b *Session) verifyAuthorshipRight(slot uint64, header *types.Header) (bool
 		return false, fmt.Errorf("could not verify signature")
 	}
 
-	// TODO: check if the producer has equivocated, ie. have they produced a conflicting block?
+	// #553 check if the producer has equivocated, ie. have they produced a conflicting block?
+	hashes, err := b.blockState.GetAllHashesForParentDepth(header)
+	if err != nil {
+		return false, err
+	}
+
+	log.Debug("Got AllHashesForParentDepth", "len(hashes)", len(hashes))
+
+	for key, v := range hashes {
+		//get header
+		currentHeader, err := b.blockState.GetHeader(key)
+		if err != nil {
+			return false, fmt.Errorf("could not verify digests, key %s, depth, %d", key, v)
+		}
+
+		//compare digest
+		currentBlockProducerIndex := getBlockProducerIndex(currentHeader)
+		existingBlockProducerIndex := babeHeader.BlockProducerIndex
+
+		log.Debug("compare Digest", "currentBlockProducerIndex", currentBlockProducerIndex,
+			"existingBlockProducerIndex", existingBlockProducerIndex)
+
+		if reflect.DeepEqual(currentBlockProducerIndex, existingBlockProducerIndex) {
+			return false, fmt.Errorf("duplicated SealDigest")
```

Comment:
I don't think reflect is needed for the comparison, but it doesn't hurt. also, the error message should say "block producer equivocated" or something like that

Reply:
ops, reflect was required before but now just a simple `==` would work as well. Also updated err to `block producer equivocated` as suggested.